File: rust/src/code/remove_duplicates_from_sorted_array.rs
struct Solution {}

impl Solution {
    pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
        let mut l: usize = 1;
        for i in 1..nums.len() {
            if nums[i] != nums[i - 1] {
                nums[l] = nums[i];
                l += 1;
            }
        }

        l as i32
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_remove_duplicates_case_1() {
        // arrange
        let mut nums: Vec<i32> = vec![1, 1, 2];
        let expected: i32 = 2;

        // act
        let actual = Solution::remove_duplicates(&mut nums);

        // assert
        assert_eq!(expected, actual);
    }
}

File: python/remove_duplicates_from_sorted_array.py
class Solution:
    def removeDuplicates(self, nums: list[int]) -> int:
        """time complexity: O(n)"""
        l: int = 1
        for r in range(1, len(nums)):
            if nums[r] != nums[r - 1]:
                nums[l] = nums[r]
                l += 1

        return l


def test_removeDuplicates_case_1():
    # arrange
    nums: list[int] = [1, 1, 2]
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.removeDuplicates(nums)

    # assert
    assert expected == actual


def test_removeDuplicates_case_2():
    # arrange
    nums: list[int] = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4]
    expected: int = 5

    # act
    solution = Solution()
    actual = solution.removeDuplicates(nums)

    # assert
    assert expected == actual
